Add how-to guides about primitives

Guillermo-Mijares-Vilarino opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Added a new how-to guide section to qiskit-terra.

Close https://github.com/Qiskit/qiskit-terra/issues/7965

### Details and comments

This new section includes 2 guides written in `.rst` format:

* Compute circuit output probabilities with ``Sampler``
* Compute an expectation value with ``Estimator``

HuangJunye commented:
@ikkoham @mrvee-qC-bee Guillermo has made two how-to guides for primitives. Can you please review them? Thanks a lot.

woodsp-ibm commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,145 @@
+###########################
+Use the Estimator primitive
```
The first sentence says about observable. I wonder if the title could be more `Compute expectation value with an Estimator`. I had often been asked how to easily compute an expectation value with qiskit (with opflow it was more work). Having this title leads more towards what they want to do and this shows how. The title otherwise does not offer much - you would already need to know you wanted to use an Estimator unless you were curious what it was for.

Similar comment for Sampler.

woodsp-ibm replied:
Yes, for me those work much better as titles.

woodsp-ibm commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,145 @@
+
+    While this guide only uses Qiskit Terra's implementation of the ``Estimator`` primitive, there are other
+    implementations of this primitive like Qiskit Terra's :class:`~qiskit.primitives.BackendEstimator`, Qiskit Aer's :class:`~qiskit_aer.primitives.Estimator`
+    and Qiskit Runtime's :class:`~qiskit_ibm_runtime.Estimator`.
```
There is also this tutorial being done around selection Qiskit/qiskit-tutorials#1385 since you mention others here. That maybe helpful to link eventually to perhaps give clarity to the fact that there other others and how to select. Now the runtime has a bunch of tutorials for the runtime primitives too - would that be a "better" link than the API ref (Or in addition?)

woodsp-ibm commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,145 @@
+
+    from qiskit.quantum_info import SparsePauliOp
+
+    obs = SparsePauliOp(["II", "XX", "YY", "ZZ"], [1, 1, -1, 1])
```
```suggestion
    obs = SparsePauliOp(["II", "XX", "YY", "ZZ"], coeffs=[1, 1, -1, 1])
```
Perhaps we can be a bit more verbose and name at least this 2nd parameter here so its easier to read/understand/know what it does.

woodsp-ibm commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,145 @@
+
+    from qiskit.primitives import Estimator
+
+    estimator = Estimator()
```
By default this gives an ideal result (akin to using statevector) - should we mention this?

Guillermo-Mijares-Vilarino replied:
I'm not sure this is the place for that. This is supposed to be only a step-by-step guide for working with an `Estimator` instead of an explanation about the differences between implementations or going into all the details of the class (this is what API reference is for).  I guess it depends on how important/useful we consider to tell people how to configure the `shots` in `options` here.

Moreover, we plan to write an explanation about the primitives soon. It might be better to cover that there. A tutorial is also a fine place to mention that.

woodsp-ibm replied:
For me since this all around an Estimator configuring it (since it can work differently) to do the job you need seems highly relevant. I would not expect any of that for the QuantumCircuit or Operator it uses but since howto for using the Estimator not saying anything I am just not sure. Though I would not say much that's for certain - which is why I said mention not explain. Part of my rationale for mentioning is that I have had to answer issues where 'its my code doesn't give the same result/fails' when going from an qiskit primitives Estimator() to an Aer Estimator() both just constructed as defaults. Knowing what it does/how its configured is key there. Since there is a note pointing to the other estimators what is the implication - they are the same and can be swapped in. I have had to give people explanation on more than one occasion. I don't think this howto should deal with that of course, but maybe it should state at least how the Estimator will behave in the config its used in.

Guillermo-Mijares-Vilarino replied:
If it's a common user pain point I guess it might be better to mention that explicitly. 

Now I'm also thinking about adding an example in which I change from the default `Statevector`-based primitives to shot-based so a user can see quickly how to set the `options` argument in the primitives, as it's not obvious if you just look at the API reference.

woodsp-ibm replied:
There are two ways there if you did not figure. It can be set when calling run()  - via kwargs (named run_options in code if you end up looking) - so you can change it on a call by call basis. The value used on run() can be also be set via the constructor - options - for all run invocations (overridden by the one if passed to run). This latter via options is more normal if a user wants to use an Estimator with an algorithm like VQE where you want as a user to config the shots used in circuit execution later when used by VQE. Since the howto is more about using an Estimator and calling run yourself maybe the shots on run() is more suitable to an example here. You can see examples of both in the unit tests here https://github.com/Qiskit/qiskit-terra/blob/fa7ac688fdce3ecc27fb7e13d8b2c3f8553b3ff8/test/python/primitives/test_estimator.py#L612-L633

woodsp-ibm commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,145 @@
+Parameterized circuits with ``Estimator``
+=========================================
+
+The :class:`~qiskit.primitives.Estimator` primitive also has the option to include unbound parameterized circuits like the one below.
```
Here this starts to get close to using an optimizer with a parameterized circuit to find minimum exp val. i.e VQE. Would we want a howto to show using an estimator for that. Of course we have a more comprehensive pre-built VQE algo - but its a nice simple howto to build a simple one. Here the parameters are applied sequentially - rather than all at once in this example.

woodsp-ibm replied:
I guess I see people using the Estimator with an optimizer to make a simple VQE. It just seemed more like an aspect someone might be wanting to do and a howto, that includes parameterized circuits, which seemed to be the aspect the latter section was bringing in, might be a task that had more relevancy. I agree its bit more complex than what is there and maybe is more applicable for a future howto. It was less about VQE and more that this a common way to use parameterized ccts with an Estimator.

woodsp-ibm commented on docs/how_to/use_estimator.rst:
```diff
@@ -46,6 +46,10 @@ Then you need to create the :class:`~qiskit.circuit.QuantumCircuit` for which yo
 
+.. note::
+
+    The :class:`~qiskit.circuit.QuantumCircuit` can't include any measurements.
```
Taken at face value this is an incorrect statement. A QuantumCircuit can include measurements. If however you are going to pass it to an Estimator, the case here, then it must not include any measurement instructions.

Guillermo-Mijares-Vilarino replied:
You're right. How about this?

> The ``QuantumCircuit`` you pass to ``Estimator`` must not include any measurement
> The ``QuantumCircuit`` you pass to ``Sampler`` must include measurements

Guillermo-Mijares-Vilarino replied:
Updated in c889e90.

HuangJunye commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,253 @@
+
+From these results you can take the expected values with the attribute :attr:`~qiskit.primitives.EstimatorResult.values`.
+
+Generally, :attr:`~qiskit.primitives.EstimatorResult.values` returns a `numpy.ndarray <https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html>`_
```
Change to use numpy intersphinx

HuangJunye commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,253 @@
+
+    exp_value = result.values[0]
+    print(exp_value)
+
```
```suggestion
```

HuangJunye commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,238 @@
+
+    quasi_dist = result.quasi_dists[0]
+    print(quasi_dist)
+
```
```suggestion
```

HuangJunye commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,238 @@
+From these results you can take the probability distributions with the attribute :attr:`~qiskit.primitives.SamplerResult.quasi_dists`.
+
+Even though there is only one circuit in this example, :attr:`~qiskit.primitives.SamplerResult.quasi_dists` returns a list of :class:`~qiskit.result.QuasiDistribution`\ s.
+Generally ``result.quasi_dists[i]`` would be the quasi-probability distribution of the ``i``-th circuit.
```
Can add a note about quasi-probability

HuangJunye commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,238 @@
+.. testoutput::
+
+    {'00': 0.4999999999999999, '11': 0.4999999999999999}
+
+
```
```suggestion
```

HuangJunye commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,238 @@
+
+    SamplerResult(quasi_dists=[{0: 0.5205078125, 3: 0.4794921875}], metadata=[{'shots': 2048}])
+
+
+
```
```suggestion
```

HuangJunye reviewed (commented):
Looks great. Thanks a lot for working on these how-to guides. I left a few small comments. Can you please address them?

1ucian0 commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+
+.. note::
+
+    While this guide only uses Qiskit Terra's implementation of the ``Sampler`` primitive, there are other
+    implementations of this primitive like Qiskit Terra's :class:`~qiskit.primitives.BackendSampler`, Qiskit Aer's :class:`~qiskit_aer.primitives.Sampler`
+    and Qiskit Runtime's :class:`~qiskit_ibm_runtime.Sampler`.
```
```suggestion
While this guide uses Qiskit’s reference implementation, the ``Sampler`` primitive can be run with any provider using :class:`~qiskit.primitives.BackendSampler` .
```

HuangJunye replied:
We are going to add link to the new providers page which will contain a section of providers that have native primitive support and the link will be http://qiskit.org/providers#native_primitives

HuangJunye replied:
Put the PR on hold until the providers page is launched, which should be in the next few days.

1ucian0 commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+Initialize the ``Sampler``
+==========================
+
+Then, you need to create a :class:`~qiskit.primitives.Sampler` object.
```
```suggestion
Then, you need to create a :class:`~qiskit.primitives.Sampler` instance.
```

Maybe you can explain here `sampler = BackendSampler(backend)` to wrap a backend that does not support it natively. I assume that if a provider implements it natively, you can `from ... import Sampler`.

ElePT commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,256 @@
+Run and get results
+===================
+
+Now that you have defined ``estimator``, you can create a :class:`~.PrimitiveJob` (subclass of :class:`~qiskit.providers.JobV1`) with the
+:meth:`~qiskit.primitives.Estimator.run` method and, then, you can get the results (as a :class:`~qiskit.primitives.EstimatorResult` object) with
+the results with the :meth:`~qiskit.providers.JobV1.result` method.
```
Similar comment as before, in my opinion you are not "creating" a job, you are calling `run` and getting a job as a result:

```suggestion
Now that you have defined your ``estimator``, you can run your estimation by calling the :meth:`~qiskit.primitives.Estimator.run` method, 
which returns an instance of :class:`~.PrimitiveJob` (subclass of :class:`~qiskit.providers.JobV1`). You can get the results from the job (as a :class:`~qiskit.primitives.EstimatorResult` object) 
with the :meth:`~qiskit.providers.JobV1.result` method.
```

ElePT commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,256 @@
+Get the expected value
+----------------------
+
+From these results you can take the expected values with the attribute :attr:`~qiskit.primitives.EstimatorResult.values`.
```
```suggestion
From these results you can extract the expected values with the attribute :attr:`~qiskit.primitives.EstimatorResult.values`.
```

ElePT commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,256 @@
+
+From these results you can take the expected values with the attribute :attr:`~qiskit.primitives.EstimatorResult.values`.
+
+Generally, :attr:`~qiskit.primitives.EstimatorResult.values` returns a :class:`numpy.ndarray`
```
It is not generally, right? It is always. 

```suggestion
:attr:`~qiskit.primitives.EstimatorResult.values` contains a :class:`numpy.ndarray`
```

ElePT commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,256 @@
+From these results you can take the expected values with the attribute :attr:`~qiskit.primitives.EstimatorResult.values`.
+
+Generally, :attr:`~qiskit.primitives.EstimatorResult.values` returns a :class:`numpy.ndarray`
+whose ``i``-th element would be the expectation value corresponding to the ``i``-th circuit and ``i``-th observable.
```
```suggestion
whose ``i``-th element is the expectation value corresponding to the ``i``-th circuit and ``i``-th observable.
```

ElePT commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,256 @@
+
+The main difference from the previous case is that now you need to include the parameter values
+for which you want to evaluate the expectation value as a ``list`` of ``list``\ s of ``float``\ s.
+The idea is that the ``i``-th element of the bigger ``list`` is the set of parameter values
```
```suggestion
The idea is that the ``i``-th element of the outer``list`` is the set of parameter values
```

ElePT commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,256 @@
+Parameterized circuits with ``Estimator``
+=========================================
+
+The :class:`~qiskit.primitives.Estimator` primitive also has the option to include unbound parameterized circuits like the one below.
```
```suggestion
The :class:`~qiskit.primitives.Estimator` primitive can be run with unbound parameterized circuits like the one below.
```

ElePT commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,256 @@
+
+It is also possible that you may want to change any other option.
+
+For example, in the previous sections the :class:`~qiskit.primitives.Estimator`
+is :class:`~qiskit.quantum_info.Statevector`-based but it can be changed
+to shot-based by setting a number of ``shots``. For reproducibility purposes, in this
+guide a ``seed`` will also be set.
```
I think that this description is a bit more accurate, all that setting shots does is adding artificial shot noise to the Statevector-based simulation:

```suggestion
By default, the reference :class:`~qiskit.primitives.Estimator` class performs an exact statevector
calculation based on the :class:`~qiskit.quantum_info.Statevector` class. However, this can be 
modified to include shot noise if the number of ``shots`` is set. 
For reproducibility purposes, a ``seed`` will also be set in the following examples.
```

ElePT commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,256 @@
+
+There are two main ways of doing this:
+
+* Setting keyword arguments in the :meth:`~qiskit.primitives.Estimator.run` method.
```
For consistency with the format of the other option:

```suggestion
* Set keyword arguments in the :meth:`~qiskit.primitives.Estimator.run` method.
```

ElePT commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+Run and get results
+===================
+
+Now that you have defined ``sampler``, you can create a :class:`~.PrimitiveJob` (subclass of :class:`~qiskit.providers.JobV1`) with the
+:meth:`~qiskit.primitives.Sampler.run` method and, then, you can get the results (as a :class:`~qiskit.primitives.SamplerResult` object) with
+the results with the :meth:`~qiskit.providers.JobV1.result` method.
```
Same as with the estimator, I don't think that "you can create a job" is obscuring what you are actually doing (which is running the sampler)

```suggestion
Now that you have defined your ``sampler``, you can run it by calling the :meth:`~qiskit.primitives.Sampler.run` method, 
which returns an instance of :class:`~.PrimitiveJob` (subclass of :class:`~qiskit.providers.JobV1`). You can get the results from the job (as a :class:`~qiskit.primitives.SamplerResult` object) 
with the :meth:`~qiskit.providers.JobV1.result` method.
```

ElePT commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+From these results you can take the probability distributions with the attribute :attr:`~qiskit.primitives.SamplerResult.quasi_dists`.
+
+Even though there is only one circuit in this example, :attr:`~qiskit.primitives.SamplerResult.quasi_dists` returns a list of :class:`~qiskit.result.QuasiDistribution`\ s.
+Generally ``result.quasi_dists[i]`` would be the quasi-probability distribution of the ``i``-th circuit.
```
```suggestion
``result.quasi_dists[i]`` is the quasi-probability distribution of the ``i``-th circuit.
```

ElePT commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+Probability distribution with binary outputs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you prefer to see the outputs as binary strings instead of decimal, you can use the
```
```suggestion
If you prefer to see the output keys as binary strings instead of decimal numbers, you can use the
```

ElePT commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+Parameterized circuits with ``Sampler``
+=========================================
+
+The :class:`~qiskit.primitives.Sampler` primitive also has the option to include unbound parameterized circuits like the one below.
+You can also bind values to the parameters of the circuit and follow the steps
+of the previous example.
```
Similar comment to the estimator:

```suggestion
The :class:`~qiskit.primitives.Sampler` primitive can be run with unbound parameterized circuits like the one below.
You can also manually bind values to the parameters of the circuit and follow the steps
of the previous example.
```

ElePT commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+    meas: 2/══════════════════╩══╩═
+                              0  1 
+
+The main difference from the previous case is that now you need to include the parameter values
```
```suggestion
The main difference from the previous case is that now you need to specify the sets of parameter values
```

ElePT commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+
+The main difference from the previous case is that now you need to include the parameter values
+for which you want to evaluate the expectation value as a ``list`` of ``list``\ s of ``float``\ s.
+The idea is that the ``i``-th element of the bigger ``list`` is the set of parameter values
```
```suggestion
The idea is that the ``i``-th element of the outer ``list`` is the set of parameter values
```

ElePT commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+Change run options
+==================
+
+It is also possible that you may want to change any other option.
```
```suggestion
Your workflow might require tuning primitive run options, such as shots.
```

ElePT commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+
+It is also possible that you may want to change any other option.
+
+For example, in the previous sections the :class:`~qiskit.primitives.Sampler`
+is :class:`~qiskit.quantum_info.Statevector`-based but it can be changed
+to shot-based by setting a number of ``shots``. For reproducibility purposes, in this
+guide a ``seed`` will also be set.
```
```suggestion
By default, the reference :class:`~qiskit.primitives.Sampler` class performs an exact statevector
calculation based on the :class:`~qiskit.quantum_info.Statevector` class. However, this can be 
modified to include shot noise if the number of ``shots`` is set. 
For reproducibility purposes, a ``seed`` will also be set in the following examples.
```

ElePT commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,239 @@
+
+There are two main ways of doing this:
+
+* Setting keyword arguments in the :meth:`~qiskit.primitives.Sampler.run` method.
```
```suggestion
* Set keyword arguments in the :meth:`~qiskit.primitives.Sampler.run` method.
```

ElePT reviewed (commented):
Hi Guillermo! Thanks for this great work, I think it will be extremely helpful for our users to have these how-tos 😄 . I have realized that in the end I added quite a lot of suggestions, but most of them are small wording suggestions (and one or two explanations I would write a bit differently, but some of this might be matter of taste), overall I think that the content is great.

Guillermo-Mijares-Vilarino commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,256 @@
+Initialize observable
+=====================
+
+The first step is to define the observable whose expected value you want to compute. This observable can be any ``BaseOperator``, like the operators from :mod:`qiskit.quantum_info`.
```
There's actually no documentation for `qiskit.quantum_info.operators.base_operator.BaseOperator` 😅 (at least that I could find)

I already tried writing ```:class:`~.BaseOperator` ```, ```:class:`~qiskit.quantum_info.operators.base_operator.BaseOperator` ``` and even ```:class:`~qiskit.quantum_info.BaseOperator` ``` but none of them generated any link.

Moreover if you check, for example, [the page of the Clifford operator](https://qiskit.org/documentation/stubs/qiskit.quantum_info.Clifford.html) you'll see that `BaseOperator` appears in "Bases" but there's no link either.

Guillermo-Mijares-Vilarino replied:
However, even though there would be no actual link if I wrote ```:class:`~.BaseOperator` ``` or  ```:class:`~qiskit.quantum_info.operators.base_operator.BaseOperator` ``` the output in the guide would look the same as if I wrote ``` ``BaseOperator`` ```. 

Would it be better if I wrote the Sphinx link in case we eventually decide to document that class?

Guillermo-Mijares-Vilarino commented:
[Providers page](http://qiskit.org/providers/#primitives) launched so I removed the "on hold" label

HuangJunye reviewed (dismissed):
LGTM. Thanks a lot for working on these how-to guides for primitives. They are going to be very useful for users learning how to use primitives.

1ucian0 commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,265 @@
+###############################################
+Compute an expectation value with ``Estimator``
+###############################################
```
I think the inclusion of the word "primitive" would make it easier to find.

```suggestion
#########################################################
Compute an expectation value with ``Estimator`` primitive
#########################################################
```

1ucian0 commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,265 @@
+    .. code-block::
+
+        from qiskit.primitives import BackendEstimator
+        from qiskit_provider import QiskitProvider
```
Maybe a way to signal that we are not referring to a literal module called like that?
```suggestion
        from <some_qiskit_provider> import QiskitProvider
```

1ucian0 commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,265 @@
+
+    observable = SparsePauliOp(["II", "XX", "YY", "ZZ"], coeffs=[1, 1, -1, 1])
+
+Initialize quantum circuits
```
Is the "circuits" plural intentional

Guillermo-Mijares-Vilarino replied:
Yes. You can actually define several circuits for one run of both `Sampler` and `Estimator`. In the case of `Estimator`  you can also pass a list of observables. 

I see that I left the "Initialize observable" title in singular though. 😅

1ucian0 replied:
and the example is with a single circuit.

1ucian0 replied:
Solved by 51a1260fea0cfa9b437af39918555af85fc520d8

1ucian0 commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,265 @@
+
+    3.999999999999999
+
+Parameterized circuits with ``Estimator``
```
Should be this the same or a different how-to?

Following https://qiskit.github.io/qiskit_sphinx_theme/tutorials/tutorials_guidelines.html#provide-concrete-steps:

> ignore options and alternatives.

Guillermo-Mijares-Vilarino replied:
This is a how-to not a tutorial. Options in how-tos are fine

1ucian0 commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,265 @@
+
+    3.999999998697238
+
+Change :class:`~qiskit.primitives.Estimator` options
```
The bullet in https://github.com/Qiskit/qiskit-terra/pull/9716/files#diff-592e455293d2eb77839a55cab187e1924e1cd6f2bf5f249c5110eaa13b5d7e84R172 says "modify". Maybe better to keep that?

```suggestion
Modify :class:`~qiskit.primitives.Estimator` options
```

1ucian0 commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,248 @@
+#####################################################
+Compute circuit output probabilities with ``Sampler``
+#####################################################
```
ditto https://github.com/Qiskit/qiskit-terra/pull/9716#discussion_r1189678219
```suggestion
###############################################################
Compute circuit output probabilities with ``Sampler`` primitive
###############################################################
```

1ucian0 commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,248 @@
+    .. code-block::
+
+        from qiskit.primitives import BackendSampler
+        from qiskit_provider import QiskitProvider
```
ditto https://github.com/Qiskit/qiskit-terra/pull/9716#discussion_r1189680518

Guillermo-Mijares-Vilarino replied:
Done in [3ee6158](https://github.com/Qiskit/qiskit-terra/pull/9716/commits/3ee61586908b0c6728b06d93d8c90d27c8431fe9)

1ucian0 commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,248 @@
+    qc.measure_all()
+    qc.draw("mpl")
+
+.. testsetup::
```
ditto https://github.com/Qiskit/qiskit-terra/pull/9716/files#r1153023515 ?

Guillermo-Mijares-Vilarino replied:
Fixed in [36d54f3](https://github.com/Qiskit/qiskit-terra/pull/9716/commits/36d54f3a625d12fbec8c9c03c4cc93005ff22be2)

1ucian0 commented on docs/how_to/use_estimator.rst:
```diff
@@ -0,0 +1,265 @@
+Change run options
+==================
+
+Your workflow might require tuning primitive run options, such as shots.
```
ditto https://github.com/Qiskit/qiskit-terra/pull/9716/files#r1189714135

Guillermo-Mijares-Vilarino replied:
Fixed in [004f4ad](https://github.com/Qiskit/qiskit-terra/pull/9716/commits/004f4ad2b73be085c45db4761d57cb18dc74f7eb)

1ucian0 commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,248 @@
+
+    SamplerResult(quasi_dists=[{0: 0.5205078125, 3: 0.4794921875}], metadata=[{'shots': 2048}])
+
+Change :class:`~qiskit.primitives.Sampler` options
```
Similar to https://github.com/Qiskit/qiskit-terra/pull/9716#discussion_r1189704507

Guillermo-Mijares-Vilarino replied:
Done in [254c8de](https://github.com/Qiskit/qiskit-terra/pull/9716/commits/254c8de822ab329082b7efd78b57377c74d2c495)

Guillermo-Mijares-Vilarino commented on docs/how_to/use_sampler.rst:
```diff
@@ -0,0 +1,248 @@
+
+    {'00': 0.4999999999999999, '11': 0.4999999999999999}
+
+Parameterized circuits with ``Sampler``
```
So you think it's better that for the first example I leave everything in singular and then use the parameterized example to explain multiple circuits (and observables in the case of `Estimator`)?

Guillermo-Mijares-Vilarino replied:
I was thinking about stating explicitly that while the current example only includes one circuit you can also pass a list of circuits

1ucian0 replied:
```suggestion
Parameterized circuit with ``Sampler``
```

1ucian0 replied:
Solved by 973f1e0e18

Guillermo-Mijares-Vilarino commented:
I added a sentence about using multiple circuits and observables when you run `Sampler` and `Estimator` and fixed the use of singular and plural when talking about defining observables and circuits.

With that all your comments should be addressed @1ucian0.

Eric-Arellano commented:
Should this be cherry-picked to Terra 0.24? We can't add it to the docs in qiskit-metapackage until then (https://github.com/Qiskit/qiskit-metapackage/issues/1746) because the metapackage builds docs off of the released version, not `main`. Cc @jlapeyre @1ucian0 @HuangJunye
